### ISSUE TITLE
Won't replace a resource when its dependency is recreated

### ISSUE THREAD

**Issue by Alfablos** [Telmate/terraform-provider-proxmox#488]
Hello,
I'm writing this terraform code to spin up a kubernetes cluster (VM) on Proxmox.

I'm using cloud-init and since there's no way to "rebuild the image" like in the WebUI I created a null_resource whose trigger is the hash of the rendered cloud-init files (user_data and network), and added it as a dependancy for the actual cluster.

```HCL
# The variable that accounts for the nodes' definition
variable "nodes" {
  default = {
    kmaster = {
      id          = 6000
      storage     = "threetb"
      description = "terraform cloud-init test"
      node        = "virt"

      boot = {
        order = "c"
        disk  = "scsi0"
      }

      net = {
        nic_type    = "virtio"
        mac_address = ""
        bridge      = "vmbr0"
      }


      vga = {
        type = "serial0"
      }

      clone_from = "kinit" # ONLY CLONE SUPPORTED by this terraform module
      # pool = "test"

      hw = {
        cores        = 4
        sockets      = 1
        RAM          = 4096
        disk_size_gb = "30G"
        disk_type = "scsi"
        disk_storage = "threetb"
        nic_type     = "virtio"
        bridge       = "vmbr0"
      }

      netinit = {
        ip          = "10.0.0.60/23"
        nameservers = ["10.0.0.11", "1.1.1.1", "8.8.8.8"]
        gw          = "10.0.0.1"
      }


      userinit = {
        snippets_path = "/mnt/mounted_disk/snippets"
        kubernetes_user = {
          name     = "kube"
          ssh_keys = ["blabla", "woofwoof"]
          visudo   = ["ALL=(ALL) NOPASSWD:ALL"]
        }
        packages = [
          "some",
          "needed",
          "packages",
          "for",
          "kubernetes"
        ]

        files = [ # Supported: content, path, encoding, owner, permissions, contentpath
          {
            content = "net.bridge.bridge-nf-call-ip6tables = 1\nnet.bridge.bridge-nf-call-iptables = 1\nciao"
            path    = "/etc/sysctl.d/k8s.conf"
          },
          {
            content  = "U0VMSU5VWD1wZXJtaXNzaXZlCg=="
            encoding = "b64"
            path     = "/etc/selinux/config"
          },
          {
            content  = "so not important"
            encoding = "b64"
            path     = "/etc/docker/daemon.json"
          },
          {
            content = "br_netfilter"
            path    = "/etc/modules-load.d/k8s.conf"
          },
          {
            contentpath = "files/kinstall_ubuntu24.sh"
            path        = "/root/kubeinstall.sh" # DO NOT TOUCH OR CHANGE IN THE USERINIT TEMPLATE ACCORDINGLY
            owner       = "root:root"
            encoding    = "b64"
            permissions = "0700"
          }
        ]
        cmds = [
          "/root/kubeinstall.sh kmaster '1.23.2-00' kube"
        ]
        reboot = true
      }
    }
  }
}


# I use some local variables to keep the code clean
locals {
  netinitvars = {
    for name, val in var.nodes : name => { ip = val.netinit.ip, nameservers = val.netinit.nameservers, gw = val.netinit.gw }
  }
  userinitvars = {
    for name, node in var.nodes : name => {
      hostname = name
      kuser          = node.userinit.kubernetes_user.name
      kuser_authorized_ssh_keys = node.userinit.kubernetes_user.ssh_keys
      kuser_visudo   = node.userinit.kubernetes_user.visudo
      packages       = node.userinit.packages
      files = node.userinit.files
      kubernetes_version = var.kversion
      commands = node.userinit.cmds
      reboot_enabled = node.userinit.reboot
    }
  }
  net_rendered = {
    for name, node in var.nodes: name => templatefile("${path.module}/files/netinit.yml.tftpl", local.netinitvars[name])
  }
  user_rendered = {
    for name, node in var.nodes: name => templatefile("${path.module}/files/userinit.yml.tftpl", local.userinitvars[name])
  }
}


resource "null_resource" "cloud_init_config_files" {        # Replace provisioners with some more IaaC method 
  for_each = var.nodes

  triggers = {
    # Any change to the rendered config files will trigger the recreation                            <================||||
    initfiles = md5("${local.net_rendered[each.key]} ${local.user_rendered[each.key]}")
  }

  connection {
    type     = "ssh"
    user     = var.pve.user
    private_key = file("${path.module}/files/id_rsa")
    host     = var.pve.host
  }

  provisioner "file" {
    content =  local.net_rendered[each.key]
    destination = "/mnt/threetb/snippets/netinit-${each.key}.yml"     # Needs fixing with the var.userinit.snippets_path
  }
  provisioner "file" {
    content = local.user_rendered[each.key]
    destination = "/mnt/threetb/snippets/userinit-${each.key}.yml"    # Needs fixing with the var.userinit.snippets_path
  }
}


resource "proxmox_vm_qemu" "cloudinit-test" {
  depends_on = [
    null_resource.cloud_init_config_files,                                          <================||||
  ]

  for_each = var.nodes

  vmid = each.value.id
  name        = each.key
  desc        = each.value.description
  target_node = each.value.node
  # The destination resource pool for the new VM
  pool = try(each.value.pool, null) != null ? each.value.pool: null

  clone = each.value.clone_from  

  cores   = each.value.hw.cores
  sockets = each.value.hw.sockets
  memory  = each.value.hw.RAM


  network {
    macaddr = try(each.value.net.mac_address, null) != null? each.value.net.mac_address: null
    model = each.value.net.nic_type
    bridge = each.value.net.bridge
}


  disk {
    size = each.value.hw.disk_size_gb
    type = each.value.hw.disk_type
    storage = each.value.hw.disk_storage
  }



  os_type   = "cloud-init"
  boot = each.value.boot.order
  bootdisk = each.value.boot.disk
  cicustom                = "user=threetb:snippets/userinit-${each.key}.yml,network=threetb:snippets/netinit-${each.key}.yml" # Fix with variable
  /* Create the Cloud-Init drive on the "local-lvm" storage */
  cloudinit_cdrom_storage = "local-lvm"



  serial {
    id =  0
    type = "socket"
  }
  vga {
    type = each.value.vga.type
  }
}


```

So I'd expect the null_resource to be destroyed, and so the proxmox_vm_qemu.
What I get is actually:
```HCL
>> terraform plan
null_resource.cloud_init_config_files["kmaster"]: Refreshing state... [id=1664071392670878091]
proxmox_vm_qemu.cloudinit-test["kmaster"]: Refreshing state... [id=virt/qemu/6000]

...

Terraform will perform the following actions:

  # null_resource.cloud_init_config_files["kmaster"] must be replaced
-/+ resource "null_resource" "cloud_init_config_files" {
      ~ id       = "1664071392670878091" -> (known after apply)
      ~ triggers = { # forces replacement
          ~ "initfiles" = "4a0736e821a62d9df43cd31307ae9e2b" -> "f46799b973de47a6e9e7bae77cecd8b2"
        }
    }

  # proxmox_vm_qemu.cloudinit-test["kmaster"] will be updated in-place
  ~ resource "proxmox_vm_qemu" "cloudinit-test" {
        id                        = "virt/qemu/6000"
        name                      = "kmaster"
        # (37 unchanged attributes hidden)

      ~ disk {
          ~ size         = "29900M" -> "30G"             # fixed by using 29900M
            # (17 unchanged attributes hidden)
        }



        # (3 unchanged blocks hidden)
    }

Plan: 1 to add, 1 to change, 1 to destroy.
```

So basically even though the null resource is going to be replaced while the vm is not.

I haven't worked with Terraform for a while now but as far as I remember the vm should definitely be replaced.
Thanks

**Comment by vrubiolo** [comment]
Interested in the outcome of this discussion as well since I see the same issue. I have not tried `depends_on` using a different provider but it looked to me that it might just be about ordering. I wished it would trigger the deletion as you pointed at though...
